### PR TITLE
Fix direct colormap

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This fixes a bug on the current main implementation of direct color mode in the Labels layer that is required for #6439 grade.

### CODE REVIEW

File: napari/_vispy/layers/image.py [hunk 1]
```diff
@@ -323,11 +323,12 @@ def downsample_texture(
     "r8": np.dtype(np.uint8),
     "r16": np.dtype(np.uint16),
     "r32f": np.dtype(np.float32),
-    None: np.dtype(np.float32),
 }
 
 _DTYPE_TO_VISPY_FORMAT = {v: k for k, v in _VISPY_FORMAT_TO_DTYPE.items()}
 
+_VISPY_FORMAT_TO_DTYPE[None] = np.dtype(np.float32)
+
```
Comment (Czaki) [inline napari/_vispy/layers/image.py]: @jni, this is a real fix. 

As without this change the `_DTYPE_TO_VISPY_FORMAT[np.float32]` return `None`, not `"r32f"` here:

https://github.com/napari/napari/blob/8ed7f4c62fa607ed1f832b140043559b0b4683d9/napari/_vispy/layers/labels.py#L581

File: napari/_vispy/layers/labels.py [hunk 2]
```diff
@@ -524,6 +524,7 @@ def _on_colormap_change(self, event=None):
                 interpolation='nearest',
             )
             self.node.shared_program['LUT_shape'] = key_texture.shape
+            self.node.shared_program['color_count'] = len(color_dict)
```
Comment (Czaki) [inline napari/_vispy/layers/labels.py:527]: this only silence openGL warning

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
@Czaki either here or in #6439 we should figure out why tests didn't catch this...

**Comment by Czaki** [comment]
@jni Test does not catch this because we do not test it :P. The #6439 contains one test for this.

File: napari/_vispy/layers/image.py (hunk 1 above)
Comment (brisvag) [reply napari/_vispy/layers/image.py]: please add a comment explaining this just above, it seems easy to overlook this in the future and move it back to the dict declaration.

Comment (Czaki) [reply napari/_vispy/layers/image.py]: Ok now?

**Review by brisvag — APPROVED** [review]

**Comment by jni** [comment]
> Test does not catch this because we do not test it :P.

Well, we have *some* tests, e.g.

https://github.com/napari/napari/blob/ef26ef34b0ddea62ba28047c7f4044d21a0db9e4/napari/layers/labels/_tests/test_labels.py#L1446-L1454

but they are bad tests, apparently. 😅 I indeed didn't find any in the vispy or qt tests.

Do you want to bring the test over to this PR @Czaki? I think it would be nice to have it here.

**Comment by jni** [comment]
@Czaki weird failures on this... 

```
  ________ ERROR at teardown of test_label_colors_matching_widget[False] _________
  
  qtbot = <pytestqt.qtbot.QtBot object at 0x7fe0f7e735d0>
  
      @pytest.fixture()
      def qt_viewer(qtbot):
          qt_viewer = QtViewer(viewer=ViewerModel())
          qtbot.add_widget(qt_viewer)
          qtbot.add_widget(qt_viewer.controls)
          qt_viewer.show()
          qt_viewer.controls.show()
          yield qt_viewer
          qt_viewer.controls.hide()
  >       qt_viewer.hide()
  E       RuntimeError: wrapped C/C++ object of type QtViewer has been deleted
  
  napari/_qt/_tests/test_qt_viewer.py:694: RuntimeError
```

**Comment by Czaki** [comment]
> @Czaki weird failures on this...

looks fixed

**Review by jni — APPROVED** [review]
Awesome!

@Czaki do you think you can explain in plain terms what the problem was with the Qt tests? I don't really follow the commit. It is probably a good investment to write a developer documentation page that explains the issues with testing Qt, cleanup failures, and how to avoid them...

This should be a later PR. Maybe I'll make an issue?

**Comment by Czaki** [comment]
> @Czaki do you think you can explain in plain terms what the problem was with the Qt tests?

No. I do not understand.

The only difference is to stop using qtbot to manage QtViewers and switch to manually manage. This was to fix leake widget warnings (as `make_napari_viewer` requires strict cleanup after tet and qtbot do this lazy). 

But I do not understand how use `qtbot` leads to `RuntimeError: wrapped C/C++ object of type QtViewer has been deleted`

**Comment by jni** [comment]
Ok. That's a bummer. 😂 

Anyway, thinking about reviewing #6459, I realise now that `selection + 1` is actually a bug: when the selection has been mapped to `n`, it will now be outside of the range of values in the texture (rather than looping back to 1), and nothing will be selected. We need to cast using the same function as the whole texture.

**Comment by Czaki** [comment]
> I realise now that `selection + 1` is actually a bug

It is an intentional bug... it is part of a bigger problem that I'm working to solve in follow-up PR. 
It is an improvement of current state as selection will start working for values below `num_colors`



> We need to cast using the same function as the whole texture.

Nope. This solution will show all components mapping to the same color. not only selected label.

**Pull request merged**